File: src/types/media.ts
/** メディアファイルの完全な情報 */
export interface MediaInfo {
  path: string;
  filename: string;
  format: FormatInfo;
  streams: StreamInfo[];
  duration: number; // 秒
  size: number; // バイト
  bitRate: number; // bps
}

/** コンテナフォーマット情報 */
export interface FormatInfo {
  name: string; // "mp4", "matroska", "avi"
  longName: string; // "QuickTime / MOV"
  formatTags: Record<string, string>;
}

/** ストリーム情報（映像/音声/字幕） */
export interface StreamInfo {
  index: number;
  streamType: StreamType;
  codecName: string;
  codecLongName: string;

  // Video specific
  width?: number;
  height?: number;
  fps?: number;
  pixFmt?: string;

  // Audio specific
  sampleRate?: number;
  channels?: number;
  channelLayout?: string;

  // Common
  bitRate?: number;
  duration?: number;
  language?: string;
  title?: string;
}

export type StreamType = 'video' | 'audio' | 'subtitle' | 'data';

/** サムネイルストリップデータ */
export interface ThumbnailStrip {
  paths: string[]; // サムネイル画像パスの配列
  interval: number; // 各サムネイルの間隔（秒）
  width: number;
  height: number;
}

/** 波形データ */
export interface WaveformData {
  samples: number[]; // -1.0 〜 1.0
  duration: number;
  sampleRate: number;
  channels: number;
}

File: src-tauri/src/models/media.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct MediaInfo {
    pub path: String,
    pub filename: String,
    pub format: FormatInfo,
    pub streams: Vec<StreamInfo>,
    pub duration: f64,
    pub size: u64,
    pub bit_rate: u64,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct FormatInfo {
    pub name: String,
    pub long_name: String,
    pub format_tags: HashMap<String, String>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct StreamInfo {
    pub index: u32,
    pub stream_type: StreamType,
    pub codec_name: String,
    pub codec_long_name: String,
    pub width: Option<u32>,
    pub height: Option<u32>,
    pub fps: Option<f64>,
    pub pix_fmt: Option<String>,
    pub sample_rate: Option<u32>,
    pub channels: Option<u32>,
    pub channel_layout: Option<String>,
    pub bit_rate: Option<u64>,
    pub duration: Option<f64>,
    pub language: Option<String>,
    pub title: Option<String>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "snake_case")]
pub enum StreamType {
    Video,
    Audio,
    Subtitle,
    Data,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct WaveformData {
    pub samples: Vec<f32>,
    pub duration: f64,
    pub sample_rate: u32,
    pub channels: u32,
}
